---
title: "AIH_Decompensation_T-Test"
output: html_document

summary: Per our discussion with Angela, I looked at the two AIH patients who had compensated cirrhosis at sample collection but who subsequently decompensated. They are AASLD-066 and -070. Then I looked for other AIH patients (matched for age, sex, cirrhosis, treatment, and length of follow-up) who did not decompensate during the follow-up period. They are AASLD-027 and -067. I didn't match for it but they all had pretty similar ALTs at the time of sample collection.

problem: Can you compare the gene expression of 66 & 70 vs. 27 & 67? You could use the method Angela suggested: a simple nonparametric test of differential gene expression based on gene counts. This could provide interesting insight into which genes are involved in decompensation, or be a first step toward predicting poor outcomes in cirrhotic AIH patients. AASLD-061 is matched for sex, cirrhosis, treatment, and follow-up time. But the age is a couple decades off and her ALT was lower than the others

---
Read in the raw gene counts table, after filtering using PriceSeqFilter and cutting TruSeq adaptors from the reads. The file is a concatenated output of STAR --quantmode, taking the first column of counts (for unstranded data) for each file.

Then we remove any genes that are ERCC genes (positive control RNA augmented to human genome in IDSeq) and water samples. We format the ENSEMBL gene names [rownames] to exclude the version number after the period. We format the file names [colnames] to extract the AASLD ID.


Load packages required, will need to install.packages("") if you do not have them already
```{r}
library("biomaRt")
library("gplots")
library("dplyr")
library("ggplot2")
library("RColorBrewer")
library("magrittr")
library("pheatmap")
library("stringr")
library("limma")
library("edgeR")
```

```{r}
genecounts<-read.delim(as.matrix("../AIH-NovaSeq-Gene-Counts/aih_full_cutfilt-85-98-90_dashed_gene_counts_final.tsv"), row.names = 1, stringsAsFactors = TRUE) 

#make an ERCC gene counts dataframe and then remove those gene counts
ERCCgenecounts <- genecounts[grep("ERCC", rownames(genecounts)), ]
genecounts <- genecounts[ !rownames(genecounts) %in% rownames(ERCCgenecounts), ]

Watergenecounts <- genecounts[,grep("water", colnames(genecounts))]
genecounts <- genecounts[,!colnames(genecounts) %in% colnames(Watergenecounts) ]

#removing top 4 rows
genecounts <- genecounts[5:nrow(genecounts),]

#take genenames from gene counts file and split them across the period, only taking the first column. this strips the version number of the ENSEMBL ID.
genenames<- sapply( strsplit( rownames(genecounts), split="\\." ), "[", 1 )
rownames(genecounts) <- genenames

#rename column names for AASLD samples 1-110, included in plate 1
colnames(genecounts) <- paste0("AASLD", "_", str_extract(colnames(genecounts), pattern = "(?<=aasld[.|_])\\d\\d\\d"))
colnames(ERCCgenecounts) <- paste0("AASLD", "_", str_extract(colnames(ERCCgenecounts), pattern = "(?<=aasld[.|_])\\d\\d\\d"))

```



Read in the metadata - Amy ran Arielle's fibrosis algorithm and metadata code with some modifications after a RedCap download by Michele (this file is called AIH_metadata_firstpass_ALedits.R). Careful to run this all as a chunk - rerunning certain parts separately will not result in the correct metadata.

We will bin continuous data in order make it compatible with DESeq2. Arielle chose correct bins based on histograms. 

```{r}


#read in metadata, rename rows so that they match gene counts file
metadata <- read.csv("../AIH-metadata/AIH_metadata_firstpass_ALedits_122018.csv", header=TRUE, row.names = 1)
metadata
#rename metadata columns to match gene counts
row.names(metadata)<-paste0("AASLD", "_", str_extract(row.names(metadata), pattern = "(?<=AASLD[.|_|-])\\d\\d\\d"))

#store the unbinned data for other purposes
metadata$age_nobin <- metadata$age
metadata$alt_nobin <- metadata$alt
metadata$igg_coll_nobin <- metadata$igg_coll

##ast summary stats and plot data; cutting into approx 25 bins (based on stats and #of bars in graph)
metadata$ast <- cut (metadata$ast, 25, labels = 1:25)

#age summary stats cutting into 3 bins (based on the numbers of bars on the graph). We can also cast age in 25 bins
metadata$age <- cut (metadata$age, 3, labels = 1:3)

#create fibrosis groups
metadata$fibrosis_groups<- ifelse(metadata$case_hl_du =="control", "healthy", ifelse(metadata$F03_F4_final == "F0_F3", "AIHnocirrhosis", "AIHcirrhosis"))

##alt summary stats cutting into three bins 
metadata$alt_groups <- ifelse(metadata$case_hl_du == "control", "healthy", ifelse(metadata$alt>30, "AIHelevatedALT", "AIHnormalALT"))

##on treatment groups
metadata$on_tx_groups <- ifelse(metadata$case_hl_du == "control", "healthy", as.character(metadata$on_tx))

##alt summary stats cutting into approx 30 bins (based on stats and #of bars in graph)
metadata$alt <- cut (metadata$alt, 30, labels = 1:30)

##response - giving healthys a 'no treatment'
metadata$response_groups <- ifelse(metadata$case_hl_du == "control", "healthy-notreat", as.character(metadata$response))

##tbili summary stats cutting into approx 15 bins (based on stats and #of bars in graph)
metadata$bili <- cut (metadata$bili, 15, labels = 1:15)

##alkp summary stats cutting into approx 20 bins (based on stats and #of bars in graph)
metadata$alkp <- cut (metadata$alkp, 20, labels = 1:20)

##igg summary stats cutting into approx 25 bins (based on stats and #of bars in graph)
metadata$igg_coll <- cut (metadata$igg_coll, 25, labels = 1:25)

#recast some variables as factors
metadata$spl_plate <- as.factor(metadata$spl_plate)
metadata$dt_isl <- as.factor(metadata$dt_isl)
metadata$response <- as.factor (metadata$response)

#remove COMPLEX CASES
metadata_filt <- metadata[!(metadata$case_hl_du == "complex" ) , ]

#remove outlier samples 67 and 94
metadata_filt <- metadata_filt[!(rownames(metadata_filt) %in% c("AASLD_067", "AASLD_094")),]

#remove patients who had a liver transplant before sample collection
metadata_filt <- metadata_filt[(is.na(metadata_filt$lt_before_coll) | metadata_filt$lt_before_coll =="no") ,]

#only look at cases with no NA - this was removed but can be added back in if needed for subgroup analyses
metadata_filt <- metadata_filt[!( is.na(metadata_filt$response_groups)), ]

#metadata_filt <- metadata_filt[!(is.na(metadata_filt$alt_groups) | is.na(metadata_filt$age) | is.na(metadata_filt$case_hl_du) | is.na(metadata_filt$sex) | is.na(metadata_filt$on_tx_groups) | is.na(metadata_filt$fibrosis_groups)| is.na(metadata_filt$race) | is.na(metadata_filt$ethn)), ]
metadata_filt <- subset(metadata_filt, rownames(metadata_filt) %in% c("AASLD_027","AASLD_061","AASLD_066","AASLD_070"))



```
```{r}
genecounts <- genecounts[,rownames(metadata_filt)]
```

We will now use biomart, a package which links gene ids from different databases, to convert ENSEMBL gene ID to HGNC symbol.
```{r eval=FALSE, include=FALSE}
#The following chunk of code uses the ENSEMBL mart, 
#querying with the ENSEMBL gene id and requesting the Entrez gene id 
#and HGNC gene symbol. 

listMarts(host="www.ensembl.org")
mart = useMart(biomart="ENSEMBL_MART_ENSEMBL",dataset="hsapiens_gene_ensembl", host="www.ensembl.org")
genemap <- getBM( attributes = c("ensembl_gene_id", "entrezgene", "hgnc_symbol", "gene_biotype", "description") ,filters = "ensembl_gene_id",
                  values = genenames,
                  mart)

```

```{r}
#create an index in which you are rearranging ensmbl gene ids from genemap into same order as they are in genenames
idx <- match(genenames, genemap$ensembl_gene_id )

#create new dataset which binds gene attributes
entrez <- genemap$entrezgene[ idx ]
hgnc_symbol <- genemap$hgnc_symbol[ idx ]
description <- genemap$description[ idx ]
gene_biotype <- genemap$gene_biotype[ idx ]
ensembl <- genemap$ensembl_gene_id[ idx ]

ga<-cbind(hgnc_symbol,description,gene_biotype,ensembl,entrez)

#make ga into a data frame from a matrix
ga<-as.data.frame(ga)
ga$gene_biotype<-as.character(ga$gene_biotype)

```

```{r}
#pc = protein coding
pc<-ga[ga$gene_biotype=="protein_coding",]
pc<-pc[!(is.na(pc$ensembl)),]

#make an index for protein coding genes
idxpc<- match(pc$ensembl,rownames(genecounts))
genecountspc<-genecounts[idxpc,]
#pc$hgnc_symbol
#rownames(genecountspc)
#rownames(genecountspc)<-make.unique(as.character(pc$ensembl))

#pc$hgnc_symbol
#length(rownames(genecountspc))
#write.csv( as.data.frame(genecountspc), file="AIH_1xPCRDASHed_NovaSeq_Plate01_nofilter_genecounts-protein-coding.csv")
```


Remove hemoglobin genes based on BioMart description (searching for word "hemoglobin"). There are 14 genes related to hemoglobin which STAR aligns to; of those, 12 are protein coding.
```{r}
#identify hemoglobin genes and write to CSV
# setwd("~/Documents/Informatics/AIH/github/AIH-Project/DESeq2-R-Scripts/")
hemoglobin <- genemap[grepl("hemoglobin[ |,]", genemap$description),]
# write.csv(genemap[grepl("hemoglobin ", genemap$description),], "hemoglobin_genes.csv")
length(rownames(genecountspc))
#rownames(genecountspc)
hemoglobin$ensembl_gene_id
#what genes are being removed? then remove them from our genecounts table
genecountspc[rownames(genecountspc) %in% hemoglobin$ensembl_gene_id, ]
genecountspc <- genecountspc[ !rownames(genecountspc) %in% hemoglobin$ensembl_gene_id, ]

#how many genes have hemoglobin? how many genes are left?
nrow(hemoglobin)
length(rownames(genecountspc))
```


Present/absent filter - have been running consistently. 
```{r}
#Current number of total genes in the set (by dimesion)
dim(genecountspc)
    
#Identify number of sammples with at least 10 reads in 20% of samples
genecountspc_filt <- apply(genecountspc, 1, function(x) ((sum(x>10)) > (0.2*ncol(genecountspc)))) 

#First pass of removing low read count genes
genecountspc_filt <- as.data.frame(genecountspc_filt)
colnames(genecountspc_filt) <- "N"

#keep only genes for which the filter is TRUE
genecountspc_filt <- genecountspc_filt[genecountspc_filt$N == "TRUE", , drop = FALSE]

dim(genecountspc_filt) # xx genes make it past the filter
FilterIdx <- match(rownames(genecountspc_filt), rownames(genecountspc))
genecountspc_filt <- genecountspc[FilterIdx,]
dim(genecountspc_filt)
```

need to take log cpm of values
```{r}
y <- DGEList(counts=genecountspc_filt)
y <- calcNormFactors(y)
log.cpm.y = cpm(y, log = TRUE)
```


```{r}
decompgroup = subset(metadata_filt, select = "decomp")
decompgroup$patientpair = c("1","2","1","2")

nocomp <- c(as.vector(log.cpm.y[,"AASLD_066"]), as.vector(log.cpm.y[,"AASLD_070"]))
yescomp <- c(as.vector(log.cpm.y[,"AASLD_027"]), as.vector(log.cpm.y[,"AASLD_061"]))

wilcox.test(nocomp,yescomp, paired = TRUE, p.adjust.methods = "BH")

#run DESeq2 on paired sample
# ~ patient.ID + treatment + response
library(DESeq2)
dds <- DESeqDataSetFromMatrix(countData = genecountspc_filt, colData = decompgroup,
                              design= ~patientpair+decomp)
dds<-DESeq(dds)

results <- results(dds, contrast=c("decomp", "yes","no"), alpha=0.05, tidy = TRUE)

results$hgnc <- genemap$hgnc_symbol[match(results$row, genemap$ensembl_gene_id)]

write.csv(results, file = "../DifferentialGeneExpression-Output/Amy_Decompensation_Analysis_010819/2019-01-09_AIH_Decompensation_Paired_Prefiltered_DEGenes_ALL.csv")
####

#genecountspc_t = t(genecountspc)
#log.cpm.y.genecountspc_t = as.data.frame(t(log.cpm.y))
#pairwise.wilcox.test(subset(log.cpm.y.genecountspc_t, select = "ENSG00000186092"), subset(metadata_filt, select = "decomp"), p.adjust.method = "BH", paired = TRUE)

```

