---
title: "AIH NovaSeq Data Preliminary Analysis"
output: html_notebook
---

This is an [R Markdown](http://rmarkdown.rstudio.com) Notebook. When you execute code within the notebook, the results appear beneath the code. 

Try executing this chunk by clicking the *Run* button within the chunk or by placing your cursor inside it and pressing *Cmd+Shift+Enter*. 

```{r}
library(DESeq2)
library(Rsamtools)
library(biomaRt)
library(GenomicAlignments)
library( gplots)
library(dplyr)
library(gplots)
library(RColorBrewer)
library(stringr)s
```



First we take the gene counts file and remove ERCC genes and select only 1XPCR DAShed samples
```{r}
#take gene_counts_final from STAR tab file, after running Caitriona's script
genecounts<-read.delim(as.matrix("../AIH NovaSeq Gene Counts/aih_full_unfiltered_dashed_gene_counts_final.tsv"), row.names = 1, stringsAsFactors = TRUE)
genecounts<-genecounts[5:nrow(genecounts),]

#make an ERCC gene counts dataframe and then remove those gene counts
ERCCgenecounts <- genecounts[grep("ERCC", rownames(genecounts)), ]
genecounts <- genecounts[ !rownames(genecounts) %in% rownames(ERCCgenecounts), ]


#take genenames from gene counts file and split them across the period, only taking the first column. this strips the version number of the ENSEMBL ID.
genenames<- sapply( strsplit( rownames(genecounts), split="\\." ), "[", 1 )
rownames(genecounts) <- genenames

```

Then we load in the metadata and filter out the duplicates, because their information is not complete and will weight the gene expression.

```{r}

#take a metadata spreadsheet with samples as first column and metadata headers as first row
metadata <- read.csv("18_0914_AASLD_AutoimmunePilot_metadata.csv", header=TRUE, stringsAsFactors = TRUE)

#filter out duplicates so we are only looking at cases and controls
metadata$case_hl_du <- as.numeric(metadata$case_hl_du)
metadatafilt <- subset(metadata, case_hl_du != 3)
rownames(metadatafilt) <- metadatafilt[,1]
metadatafilt <- metadatafilt[,-1]
metadatafilt[colnames(metadatafilt)]<-lapply(metadatafilt[colnames(metadatafilt)], factor)

#extract only the aasld id from the long filename
rownames(metadatafilt)
rownames(metadatafilt)<-str_extract(rownames(metadatafilt), pattern = "aasld[.|_]\\d+")
colnames(genecounts) <- str_extract(colnames(genecounts), pattern = "aasld[.|_]\\d+")

#make genecounts table that has only samples in rownames table
genecountsfilt <- genecounts [colnames(genecounts) %in% rownames(metadatafilt)]
colnames(genecountsfilt)
rownames(metadatafilt)
metadatafilt <- metadatafilt [rownames(metadatafilt) %in% colnames(genecountsfilt),]
```

Confirm that dataframes are compatible with DESeq2

```{r}

#check that column names of gene counts equals rownames of gene counts
all(rownames(metadatafilt)==colnames(genecountsfilt))

```

Use biomart to pull gene attributes that can be linked to the ENSEMBL ID and create a new data frame that includes these attributes

```{r}

#The following chunk of code uses the ENSEMBL mart, 
#querying with the ENSEMBL gene id and requesting the Entrez gene id 
#and HGNC gene symbol. 

#ensemble use HG38 as a reference
listMarts(host="www.ensembl.org")
mart = useMart(biomart="ENSEMBL_MART_ENSEMBL",dataset="hsapiens_gene_ensembl", host="www.ensembl.org")
genemap <- getBM( attributes = c("ensembl_gene_id", "entrezgene", "hgnc_symbol", "gene_biotype", "description"),
                  filters = "ensembl_gene_id",
                  values = genenames,
                  mart)

#create an index in which you are rearranging ensmbl gene ids from genemap into same order as they are in genenames
idx <- match(genenames, genemap$ensembl_gene_id )

#create new dataset which binds gene attributes
entrez <- genemap$entrezgene[ idx ]
hgnc_symbol <- genemap$hgnc_symbol[ idx ]
description <- genemap$description[ idx ]
gene_biotype <- genemap$gene_biotype[ idx ]
ensembl <- genemap$ensembl_gene_id[ idx ]

ga<-cbind(hgnc_symbol,description,gene_biotype,ensembl,entrez)

#make ga into a data frame from a matrix
ga<-as.data.frame(ga)
ga$gene_biotype<-as.character(ga$gene_biotype)

```

Create a new dataframe that encludes only the protein coding genes and gene counts

```{r}
#pc = protein coding
pc<-ga[ga$gene_biotype=="protein_coding",]
pc<-pc[!(is.na(pc$ensembl)),]

#make an index for protein coding genes
idxpc<- match(pc$ensembl,rownames(genecountsfilt))
genecountspc<-genecountsfilt[idxpc,]
rownames(genecountspc)<-make.unique(as.character(pc$hgnc_symbol))

write.csv( as.data.frame(genecountspc), file="AIH_1xPCRDASHed_NovaSeq_full_cutfilt_genecounts-protein-coding.csv")
```

We can now answer the question, how many of our gene counts are considered protein coding?

```{r}
#find the total number of gene counts to HG38 for each sample
genecountstotal<-cbind(total = colSums(genecountsfilt), row.names(colnames(genecountsfilt)))
#find the number of protein coding genes for each sample
genecountstotal <- as.data.frame(cbind(genecountstotal, pc=colSums(genecountspc)))
#calculate the percentage
genecountstotal <- mutate(genecountstotal, percentpc = pc/total)
#View the results
head(genecountstotal)
```

Before running through DESeq2, we may want to do some filtering of low count genes, which contribute to noise in our data. Here is an example pre-filter step from Stephanie:
```{r eval=FALSE, include=FALSE}
#PREFILTERING: keep genes with at least 10 counts in 20% of samples
test <- data.frame(sample1=c(20,3,10,1),sample2=c(4,10,11,2), sample3=c(0,0,2,1), sample4=c(2,1,11,10),sample5=c(0,0,0,0))
mutate(test, atleast10 = apply(test, 1, function(x) ((sum(x>10)) > (0.2*ncol(test)))) )

genecountspc_filtered <- as.data.frame(mutate(genecountspc, atleast10 = apply(genecountspc, 1, function(x) ((sum(x>10)) > (0.2*ncol(genecountspc)))) ))
genecountspc_filtered <- as.data.frame(filter(genecountspc_filtered, genecountspc_filtered$atleast10 == TRUE))
genecountspc_filtered <- subset(genecountspc_filtered, select = -atleast10)
nrow(genecountspc) - nrow(genecountspc_filtered)
```

We can now run DESeq2 analysis. Make sure you have selected the gene counts table you would like to use (protein coding, filtered, etc) and then use the metadata columns to choose your design variables.

```{r}
#view potential options for the design
colnames(metadatafilt)
#create DESeq object
dds<-DESeqDataSetFromMatrix(countData= genecountspc, colData= metadatafilt, design=~sex+case_hl_du )
#perform standard DESeq2 analysis
dds<-DESeq(dds)
```

Select a variable you would like to contrast, and generate statistics on it using DESeq2's results function
```{r}
#generate results based on diagnosis of patients
resDiag<-results(dds, contrast=c("case_hl_du", "2", "1"), alpha=0.2)
#write a table with the results
write.table(resDiag, "AIH_Diagnosis_full_cutfilt_DESeq2_Results.tsv", sep="\t")
```

You can now query the results dataframe for some statistics

```{r}
#how many of the results have a P-value of less than 0.05?
sum( resDiag  $pvalue < 0.05, na.rm=TRUE ) 

#how many genes at fdr<0.2 ?
sum( resDiag$padj < 0.2, na.rm=TRUE ) 
```
Create tables with all genes below a certain FDR, and write a CSV with this. Also, can write a csv with the normalized counts.

```{r}
#Reorder results by log2foldchange
resDiag<- resDiag[ order( resDiag$log2FoldChange ), ] 


#pull out just FDR<0.2
resDiag_subset <- resDiag[ which(resDiag$padj < 0.2), ]
write.csv(resDiag_subset, file="AIHvControl_full_sexinc_cutfilt_FDR0.2_Genes.csv")

#returns the normalized DE gene counts
#corrected for size factors
normcounts<-counts(dds, normalized = TRUE)
rownames(normcounts)<-rownames(genecountspc)
colnames(normcounts)<-colnames(genecountspc)
write.csv( as.data.frame(normcounts), file="NormCounts_full_sexinc_cutfilt_AIH-v-healthy.csv" )
```

We can now variance stablize the data for plotting. Another option would be rlog, which is much slower but more accurate.

```{r}
?vst
vst <- vst(dds, blind=TRUE)
vstassay <- assay(vst)

#rld <- rlog( dds )
#rldassay<-assay(rld)
#rownames(rldassay)<-rownames(genecountspc)
#colnames(rldassay)<-colnames(genecountspc)

```

Clear the plotting area
```{r}
install.packages("devtools")
library(devtools)
install_github("vqv/ggbiplot")
library(ggbiplot)
genecountspc

#any variable you would like to examine must be included in intgroup
?plotPCA
data <- plotPCA(vst, intgroup=c("case_hl_du","sex"), returnData=TRUE)
percentVar <- round(100 * attr(data, "percentVar"))
#can change color, shape, fill, etc to see multiple labels at a time
ggplot(data, aes(as.numeric(PC1), as.numeric(PC2), color=case_hl_du, shape = sex)) +
  geom_point(size=3) +
  geom_text(aes(label=""),hjust=.5, vjust=0)+
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance"))
```


Create a heat map of the top 30 genes based on mean, using variance stabalized data

```{r}

#take the normcounts from dds numbers 1:30, starting with the highest 30
select <- order(rowMeans(counts(dds,normalized=TRUE)),decreasing=TRUE)[1:30]

#set the color palette 
hmcol <- colorRampPalette(brewer.pal(9, "GnBu"))(100)

#this just gives you the highest base means aross samples (quality plot)

heatmap.2(counts(dds,normalized=TRUE)[select,], col = hmcol,
          Rowv = FALSE, Colv = FALSE, scale="none",
          dendrogram="none", trace="none", margin=c(3,3))

```

Unsupervised distance plot

```{r}
# dist = compute the distances between the rows of a data matrix
distsRL <- dist(t(vst)) # t = transpose a rectangular object
mat <- as.matrix(distsRL)
colData(dds)

#you can change the variable that you label the graph with (ie, sex, AIH)
paste(dds$sex)
rownames(mat) <- colnames(mat) <- with(colData(dds),
                                       paste(sex))
hc <- hclust(distsRL)
heatmap.2(mat, Rowv=as.dendrogram(hc),
          symm=TRUE, trace="none",
          col = rev(hmcol), margin=c(5, 5))

```

PCA plot
```{r}
library(ggplot2)
#any variable you would like to examine must be included in intgroup
data <- plotPCA(vst(dds), intgroup=c("case_hl_du","sex", "race", "ethn"), returnData=TRUE, blind=TRUE)
?plotPCA
percentVar <- round(100 * attr(data, "percentVar"))
#can change color, shape, fill, etc to see multiple labels at a time
ggplot(data, aes(as.numeric(PC1), as.numeric(PC2), color=case_hl_du, shape = sex)) +
  geom_point(size=3) +
  geom_text(aes(label=""),hjust=.5, vjust=0)+
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance"))
```

Heat map of signifigant genes with a dendrogram to show clustering

```{r}
#SC makes color palettes for heat maps this way
brewer=rgb(colorRamp(c ("blue", "white", "red"), space="rgb", interpolate="linear")(0:255/255), maxColorValue=255)

#find these rownames from the results in the normalized counts file (rldassay) with the sig you want and create an index
sigind<-match(rownames(resDiag_subset), rownames(vst))

#this creqtes the data object, selecting the normalized counts at the sig you want
set<-vst[sigind,]
head(set)

#Top Table heatmap without colorbars 
#t(set) transposes to allow for centering and scaling
sett<-t(set)
sets<-scale(sett, center=T, scale=T)                   
setst<-t(sets)
rm(sett)                   
brewer=rgb(colorRamp(c("darkblue", "blue3", "blue2",  'blue', "white",  "red", "red3", "red4", "darkred"), space="rgb", interpolate="linear")(0:255/255), maxColorValue=255)
#brewer=rgb(colorRamp(c ("blue", "white", "darkred"), space="rgb", interpolate="linear")(0:255/255), maxColorValue=255)

#Colorbar for AIH Status
status<- dds$case_hl_du
f.status<- factor(status)
vec.status<- rainbow(nlevels(f.status),start=.6,end=0)
status.colour <- rep(0,length(f.status))
vec.status[f.status[1]==levels(f.status)]

#creates different colors for diagnosis/no diagnosis heatmap
for(i in 1:length(f.status))
  status.colour[i] <- vec.status[ f.status[i]==levels(f.status) ]

# Make Heatmap 
heat.c<-heatmap.2(setst, col = brewer, cexRow=0.2, ColSideColors=status.colour, srtCol=45, scale="none", symbreaks=T, trace="none", margin=c(20,10))

```



