---
title: "AIH_DESeq2"
author: "AIH team (Amy, Arielle)""
date: "August, 2018"
output: html_document
editor_options: 
  chunk_output_type: console
---

Getting started with the correct files and libraries 
##Set working director and load in packages

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "")

library("DESeq2")
library("Rsamtools")
library("biomaRt")
library("GenomicAlignments")
library("gplots")
library("dplyr")
library("ggplot2")
library("RColorBrewer")
library("magrittr")
library("pheatmap")
library(stringr)

```


First we take the gene counts file and remove ERCC genes and select only 1XPCR DAShed samples
##Adding in new column names to match the formatted row names in the updated AIH-metadata file

```{r}

setwd("~/Documents/Informatics/AIH/github/AIH-Project/AIH NovaSeq Gene Counts/")
#take gene_counts_final from STAR tab file, after running Caitriona's script
genecounts<-read.delim(as.matrix("../AIH NovaSeq Gene Counts/aih_full_cutfilt-85-98-90_dashed_gene_counts_final.tsv"), row.names = 1, stringsAsFactors = TRUE) 

#make an ERCC gene counts dataframe and then remove those gene counts
ERCCgenecounts <- genecounts[grep("ERCC", rownames(genecounts)), ]
genecounts <- genecounts[ !rownames(genecounts) %in% rownames(ERCCgenecounts), ]

Watergenecounts <- genecounts[,grep("water", colnames(genecounts))]
genecounts <- genecounts[,!colnames(genecounts) %in% colnames(Watergenecounts) ]

#removing extra row at the bottom
colnames(genecounts)
genecounts <- genecounts[,-(111)]
genecounts <- genecounts[5:nrow(genecounts),]

#take genenames from gene counts file and split them across the period, only taking the first column. this strips the version number of the ENSEMBL ID.
genenames<- sapply( strsplit( rownames(genecounts), split="\\." ), "[", 1 )
rownames(genecounts) <- genenames

#rename column names for AASLD samples 1-110, included in plate 1
colnames(genecounts) <- paste0("AASLD", "_", str_extract(colnames(genecounts), pattern = "(?<=aasld[.|_])\\d\\d\\d"))

#clean up environment
rm (ERCCgenecounts)

```



Reading in the metadata
###Working directory is re-set for the latest metadata; here the column data string types should be factors (given that this is required for inclusion in DESeq2), and dupicates should have full clinical information
###Then we check to be sure rowdata == coldata 

```{r}

#read in metadata, remaname rows so that they match gene counts file
setwd("~/Documents/Informatics/AIH/github/AIH-Project/DESeq2-R-Scripts/")
metadata <- read.csv("AIH_metadata_firstpass_cut.csv", header=TRUE, row.names = 1)
#removing water control
metadata <- metadata[-c(56),]
#rename metadata columns to match gene counts
row.names(metadata)<-paste0("AASLD", "_", str_extract(row.names(metadata), pattern = "(?<=AASLD[.|_|-])\\d\\d\\d"))

##recast all integers as factor data (taking continous data and binning it)

##ast summary stats and plot data; cutting into approx 25 bins (based on stats and #of bars in graph)
metadata$ast <- cut (metadata$ast, 25, labels = 1:25)

#age summary stats cutting into approximately 20 bins (based on the numbers of bars on the graph)
metadata$age <- cut (metadata$age, 20, labels = 1:20)

##alt summary stats cutting into approx 30 bins (based on stats and #of bars in graph)
metadata$alt <- cut (metadata$alt, 30, labels = 1:30)

##tbili summary stats cutting into approx 15 bins (based on stats and #of bars in graph)
metadata$bili <- cut (metadata$bili, 15, labels = 1:15)

##alkp summary stats cutting into approx 20 bins (based on stats and #of bars in graph)
metadata$alkp <- cut (metadata$alkp, 20, labels = 1:20)

##igg summary stats cutting into approx 25 bins (based on stats and #of bars in graph)
metadata$igg_coll <- cut (metadata$igg_coll, 25, labels = 1:25)

metadata$response <- as.factor (metadata$response)

```


Confirming that row names == column names 
##Before running deseq, you will need to have a data frame with the samples as columns, and the genes as rows, with the raw counts as entries in the data frame (aka genecounts) You will also need a data frame with the samples as rows, and any metadata as columns (aka metadata, the clinical variables to use for differential expression or adjustments, information on match). The columns of the countData must match the rows of colData.  

```{r}
#Check to ensure the count dat and metadata are in the same order (result = TRUE)
all(rownames(metadata) == colnames(genecounts))

```

```{r}
#The following chunk of code uses the ENSEMBL mart, 
#querying with the ENSEMBL gene id and requesting the Entrez gene id 
#and HGNC gene symbol. 

listMarts(host="www.ensembl.org")
mart = useMart(biomart="ENSEMBL_MART_ENSEMBL",dataset="hsapiens_gene_ensembl", host="www.ensembl.org", ensemblRedirect = FALSE)
genemap <- getBM( attributes = c("ensembl_gene_id", "entrezgene", "hgnc_symbol", "gene_biotype", "description"),
                  filters = "ensembl_gene_id",
                  values = genenames,
                  mart)
```

Use biomart to pull gene attributes that can be linked to the ENSEMBL ID and create a new data frame that includes these attributes

```{r}
#create an index in which you are rearranging ensmbl gene ids from genemap into same order as they are in genenames
idx <- match(genenames, genemap$ensembl_gene_id )

#create new dataset which binds gene attributes
entrez <- genemap$entrezgene[ idx ]
hgnc_symbol <- genemap$hgnc_symbol[ idx ]
description <- genemap$description[ idx ]
gene_biotype <- genemap$gene_biotype[ idx ]
ensembl <- genemap$ensembl_gene_id[ idx ]

ga<-cbind(hgnc_symbol,description,gene_biotype,ensembl,entrez)

#make ga into a data frame from a matrix
ga<-as.data.frame(ga)
ga$gene_biotype<-as.character(ga$gene_biotype)

```

Create a new dataframe that includes only the protein coding genes and gene counts

```{r}
#pc = protein coding
pc<-ga[ga$gene_biotype=="protein_coding",]
pc<-pc[!(is.na(pc$ensembl)),]

#make an index for protein coding genes
idxpc<- match(pc$ensembl,rownames(genecounts))
genecountspc<-genecounts[idxpc,]
rownames(genecountspc)<-make.unique(as.character(pc$hgnc_symbol))

#write.csv( as.data.frame(genecountspc), file="AIH_1xPCRDASHed_NovaSeq_Plate01_nofilter_genecounts-protein-coding.csv")
```

We can now answer the question, how many of our gene counts are considered protein coding?

```{r}

#find the total number of gene counts to HG38 for each sample
genecountstotal<-cbind(total = colSums(genecounts), row.names(colnames(genecounts)))
#find the number of protein coding genes for each sample
genecountstotal <- as.data.frame(cbind(genecountstotal, pc=colSums(genecountspc)))
#calculate the percentage
genecountstotal <- mutate(genecountstotal, percentpc = pc/total)
#View the results
head(genecountstotal)

#clean environment
#rm (ga, genemap, mart, description, ensembl, entrez, gene_biotype, hgnc_symbol, idx)

```


Quality filtering
##We have to decide on an approach to quality filtering; both of the total data by PCA as well as removing low read count genes
##Example using gene-level PCA to identify outlier genes not relevent to analysis; SC's group also removed low-quality samples by outliers/low quality reads

```{r}

#SC removed low quality samples based on outliers by PCA and lowest raw read counts, which they performed in QA prior to running any analyses -- these are not evident from plate 1 of our data

##We can remove highly expressed genes that don't contribute much to variability, which was separatenly identified in PCA plotting code
###consider re-adding 6552, 6186 depending on results table
###second step removal ([-c(6999, 10121, 13826, 13827),] = 317 results <0.1, 89  results <0.05 by padj if included; with no filtering steps, 340 results <0.1, 85 signficant results <0.05)

genecountspc_gene_filt <- genecountspc
genecountspc_gene_filt_4 <- genecountspc [-c(6999, 10121, 13826, 13827),]

#Check to ensure the count dat and metadata are in the same order
all(rownames(metadata) == colnames(genecountspc_gene_filt))

# ID  samples with missing information that will be used for adjustments in model

which(is.na(metadata$case_hl_du)) # samples 18 and 22 do not have case/control designations as they are "complex cases"
which(is.na(metadata$igg) == TRUE)  #other key variables are complete, can check per left

#removing complex cases
metadata_filt <- metadata[-c(18,22),]

#write this file for downstream app use
write.csv(metadata_filt, file = "AIH_metadata_filt.csv")


genecountspc_gene_filt2 <- genecountspc_gene_filt [ ,-c(18,22),]
genecountspc_gene_filt_4 <- genecountspc_gene_filt_4 [ ,-c(18,22),]

#Check to ensure the count dat and metadata are in the same order
all(rownames(metadata_filt) == colnames(genecountspc_gene_filt2))
all(rownames(metadata_filt) == colnames(genecountspc_gene_filt_4))

# Clean up environment (TBD)
# rm (genecounts, genecountspc, metadata, genecountstotal, pc, genenames, idxpc)

```



###Remove low count genes (at least 10 counts in 20% of samples - adapted from SC lab code, who set this criteria internally)
```{r}

#Current number of total genes in the set (by dimesion)
dim(genecountspc_gene_filt2) #13348
    
#Identify number of sammples with at least 10 reads in 20% of samples
Filter <- apply(genecountspc_gene_filt2, 1, function(x) ((sum(x>10)) > (0.2*ncol(genecountspc_gene_filt2)))) # list of genes with number of samples that have at least 10 reads

Filter4 <- apply(genecountspc_gene_filt_4, 1, function(x) ((sum(x>10)) > (0.2*ncol(genecountspc_gene_filt_4)))) # list of genes with number of samples that have at least 10 reads from filtered data

#First pass of removing low read count genes
Filter <- as.data.frame(Filter)
colnames(Filter) <- "N"
Filter <- Filter[Filter$N == "TRUE", , drop = FALSE]
dim(Filter) # xx genes make it past the filter
FilterIdx <- match(rownames(Filter), rownames(genecountspc_gene_filt2))
genecountspc_gene_filt2 <- genecountspc_gene_filt2[FilterIdx,]

#Second pass -- repeat removing low read count genes, repeated wtih filtered data
Filter4 <- as.data.frame(Filter4)
colnames(Filter4) <- "N"
Filter4 <- Filter4[Filter4$N == "TRUE", , drop = FALSE]
dim(Filter4) # xx genes make it past the filter
FilterIdx4 <- match(rownames(Filter4), rownames(genecountspc_gene_filt_4))
genecountspc_gene_filt_4 <- genecountspc_gene_filt_4[FilterIdx4,]

dim(genecountspc_gene_filt2)
dim(genecountspc_gene_filt_4)

#Check to ensure the count dat and metadata are in the same order
all(rownames(metadata_filt) == colnames(genecountspc_gene_filt2))


#rm(Filter, FilterIdx, Filter4, FilterIdx4) #clean environment

```

##Preparing metadata / Running DESeq 

``` {r, error=FALSE, warning=FALSE, message=FALSE} 

#Relevel metadta for appropriate reference objects 
metadata_filt$case_hl_du <- relevel (metadata_filt$case_hl_du, ref = "control")
#metadata_filt$case_hl_du <- relevel (metadata_filt$F03_F4, ref = "F0")


#create DESeq object
dds <- DESeqDataSetFromMatrix(countData = genecountspc_gene_filt2, colData = metadata_filt, design= ~ plate + case_hl_du)

dds_filt_4 <- DESeqDataSetFromMatrix(countData = genecountspc_gene_filt_4, colData = metadata_filt, design= ~ plate + case_hl_du)

dds$sample
colnames(dds)
#rename duplicates in order to collapse replicates by sample name
#Note list of duplicate pairs (9/72, 19/11, 29/74, 39/96, 49/27, 59/75, 69/26, 79/32, 89/61, 99/23)
dds$sample <- colnames(dds)

dds$sample[69] <- "AASLD_009"
dds$sample[11] <- "AASLD_019"
dds$sample[71] <- "AASLD_029"
dds$sample[92] <- "AASLD_039"
dds$sample[25] <- "AASLD_049"
dds$sample[72] <- "AASLD_059"
dds$sample[24] <- "AASLD_069"
dds$sample[30] <- "AASLD_079"
dds$sample[59] <- "AASLD_089"
dds$sample[21] <- "AASLD_099"

?collapseReplicates
#collapse replicates
ddsCollapsed <- collapseReplicates( dds,
groupby = dds$sample)
head( as.data.frame( colData(ddsCollapsed)[ ,"sample"] ), 110)

#Verify summing has peen correctly performed
original <- rowSums( counts(dds)[ , dds$sample == "AASLD_019" ] )
all( original == counts(ddsCollapsed)[ ,"AASLD_019" ] )

#rename duplicates in order to collapse replicates by sample name, repeating for DESeq Filt4

dds_filt_4$sample <- colnames(dds)

dds_filt_4$sample[69] <- "AASLD_009"
dds_filt_4$sample[11] <- "AASLD_019"
dds_filt_4$sample[71] <- "AASLD_029"
dds_filt_4$sample[92] <- "AASLD_039"
dds_filt_4$sample[25] <- "AASLD_049"
dds_filt_4$sample[72] <- "AASLD_059"
dds_filt_4$sample[24] <- "AASLD_069"
dds_filt_4$sample[30] <- "AASLD_079"
dds_filt_4$sample[59] <- "AASLD_089"
dds_filt_4$sample[21] <- "AASLD_099"

#collapse replicates
ddsCollapsed_filt_4 <- collapseReplicates( dds_filt_4,
groupby = dds_filt_4$sample)
head( as.data.frame( colData(ddsCollapsed_filt_4)[ ,"sample"] ), 110)

#Verify summing has peen correctly performed
original <- rowSums( counts(dds_filt_4)[ , dds_filt_4$sample == "AASLD_079" ] )
all( original == counts(ddsCollapsed_filt_4)[ ,"AASLD_079" ] )

#perform standard DESeq2 analysis
dds<-DESeq(ddsCollapsed)

#repeat analysis with filtered genes
dds_filt_4 <-DESeq(ddsCollapsed_filt_4)

#example from SC's data

#dds <- DESeqDataSetFromMatrix(countData = PCounts, colData = colData, design = ~ COPDtxt + plate + CURRENT_SMOKER_V1) 
#dds<-DESeq(dds)

```

### AIH cases vs controls - building the results table

```{r, error=FALSE, warning=FALSE, message=FALSE}

#generate results using case vs control as the contrast, order by adjusted p-value
results_plate1 <- results(dds, contrast=c("case_hl_du", "case", "control"), alpha=0.05, tidy = TRUE)
results_plate1_significant <- results_plate1 %>% filter (results_plate1$padj < 0.05) %>% arrange (padj, log2FoldChange)

dim(results_plate1_significant)

results_plate1_filt_4 <- results(dds_filt_4, contrast=c("case_hl_du", "case", "control"), alpha=0.05, tidy = TRUE)
results_plate1_filt_4_significant <- results_plate1_filt_4 %>% filter (results_plate1_filt_4$padj < 0.05) %>% arrange (padj, (log2FoldChange))

dim(results_plate1_filt_4_significant)

#Checking compairson of filters vs no filter results
results_plate1_matching <- inner_join(results_plate1_significant, results_plate1_filt_4_significant, by = "row")

#MA plot of signficant reuslts (red) vs all genes to assess overall results
results_plate1_filt4sig_plotMA <- results(dds, contrast=c("case_hl_du", "case", "control"), alpha=0.05)
plotMA (results_plate1_filt4sig_plotMA, ylim = c(-1, 1))


#MA plot of signficant reuslts (red) vs all genes to assess overall results
results_plate1_filt4sig_plotMA <- results(dds_filt_4, contrast=c("case_hl_du", "case", "control"), alpha=0.05)
plotMA (results_plate1_filt4sig_plotMA, ylim = c(-1, 1))

# Make a basic volcano plot
with(results_plate1_filt4sig_plotMA, plot(log2FoldChange, -log10(pvalue), pch=20, main="Volcano plot, Plates 1&2", xlim=c(-2.5,2.5)))


# Add colored points onto above volcano plot: red if padj<0.05, gray of log2FC>1, red if both
with(subset(results_plate1_filt4sig_plotMA, padj<.05 ), points(log2FoldChange, -log10(pvalue), pch=20, col="gray"))
with(subset(results_plate1_filt4sig_plotMA, abs(log2FoldChange)>1), points(log2FoldChange, -log10(pvalue), pch=20, col="blue"))
with(subset(results_plate1_filt4sig_plotMA, padj<.05 & abs(log2FoldChange)>1), points(log2FoldChange, -log10(pvalue), pch=20, col="red"))

plot()

#write a table with the results
View(results_plate1_filt_4_significant)
write.table(results_plate1_filt_4_significant, "AIH_filt_4_bydiagnosis_DESeq2_Results~plate.tsv", sep="\t", quote = FALSE, row.names = FALSE)


#returns the normalized DE gene counts
#corrected for size factors
ddsCollapsed<-estimateSizeFactors(ddsCollapsed)
normcounts<-counts(ddsCollapsed, normalized = TRUE)
rownames(normcounts)<-rownames(genecountspc)
colnames(normcounts)<-colnames(genecountspc)
write.csv( as.data.frame(normcounts), file="NormCounts_full_cutfilt_AIH-v-healthy_~plate.csv" )


```



#For other downstream analyses, such as visualization and clustering, we transform the data (eg RLE count for PCA Or other applications, eg IPA, and VSD for WGCNA)

```{r}

#These are two ways to transform the data - rlog is more robust, but takes much longer to run (2-3 hours); saved as a piece of R data
rld <- rlog(dds_filt_4, blind=FALSE) 
save(rld, file = "RLENormalized_AIH_filt_4_bydiagnosis.RData")

##Will generate variance stabilizing transformation for use in WGNCA, saved as a piece of R data
vsd <- varianceStabilizingTransformation(dds_filt_4, blind=FALSE) 
save(vsd, file = "VSDNormalized_AIH_filt_4_bydiagnosis.RData")
?varianceStabilizingTransformation
#Creatine an output file for use in WGCNA and other downstream applications
matrix_vsd <- assay(vsd)
t_matrix_vsd <- t(matrix_vsd)
write.csv (t_matrix_vsd, file = "VSDNormalized_AIH_filt_4_bydiagnosis.csv")

```

#PCA of the data
Playing with DESeq PCA functions 
```{r}
#PlotPCA
pca_dds_filt_4 <- plotPCA (rld, intgroup=c ("case_hl_du", "sex", "race", "ethn", "aih_type", "on_tx", "igg_coll", "response", "relapse", "decomp", "lt", "liver_death", "age", "F03_F4", "alt", "ast", "bili", "alkp", "dup", "F03_F4"), returnData=TRUE)
pca_perc_var_filt_4 <- round(100 * attr(pca_dds_filt_4 , "percentVar"))

#Using PCA for ggplot2 analysis 
#Sort by duplicates -- #Note list of duplicate pairs (9/72, 19/11, 29/74, 39/96, 49/27, 59/75, 69/27, 79/32, 89/61, 99/23 )
ggplot(pca_dds_filt_4, aes(PC1, PC2, color= dup)) +
  geom_text(size =3, aes(label = name)) +
  xlab(paste0("PC1: ",pca_perc_var_filt_4[1],"% variance")) +
  ylab(paste0("PC2: ",pca_perc_var_filt_4[2],"% variance")) + 
  coord_fixed()

#Using PCA for ggplot2 analysis 
#Sort by duplicates -- #Note list of duplicate pairs (9/72, 19/11, 29/74, 39/96, 49/27)
ggplot(pca_dds_filt_4, aes(PC1, PC2, color=F03_F4)) +
  geom_text(size =3, aes(label = name)) +
  xlab(paste0("PC1: ",pca_perc_var_filt_4[1],"% variance")) +
  ylab(paste0("PC2: ",pca_perc_var_filt_4[2],"% variance")) + 
  coord_fixed()


```


#Heatmap per DESeq protocols using Sample Distance --> Matrix
```{r}
library(RColorBrewer)
#Calculating sample distances and then making a matrix of outputs of sample distances 
sampleDist_dds_filt_4 <- dist(t(assay(vst)))
sampleDistMatrix_filt_4 <- as.matrix(sampleDist_dds_filt_4)

#Creating a heat map with blue colors (using RColorBrewer) and creating heatmap with pheatmap package
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix_filt_4,
         clustering_distance_rows=sampleDist_dds_filt_4,
         clustering_distance_cols=sampleDist_dds_filt_4,
         col=colors)

```





